usim/py/core.py

```diff
@@ -0,0 +1,321 @@
+"""
+This module provides the :py:class:`~usim.py.core.Environment`,
+which hosts any SimPy simulation. In order to create new events,
+components of the simulation need access to the current environment
+- it is common to pass this as the parameter ``env``.
+
+An environment can be run standalone, or embedded into a μSim simulation.
+The latter allows interactions between the μSim and SimPy components.
+"""
+from typing import Optional, List, Tuple, Coroutine, Generator, TypeVar, Iterable,\
+    Union
+from .._core.loop import __LOOP_STATE__, Loop
+from .. import time, run as usim_run, Concurrent
+from .. import Scope
+
+from .events import Event
+from .exceptions import NotCompatibleError, StopSimulation, StopProcess
+
+
+class EnvironmentScope(Scope):
+    def _is_suppressed(self, exc_val):
+        return isinstance(exc_val, StopSimulation) or super()._is_suppressed(exc_val)
+
+    PROMOTE_CONCURRENT = Scope.PROMOTE_CONCURRENT + (StopSimulation,)
+
+
+def _inside_usim():
+    """Whether the current stack is run by ``usim``"""
+    try:
+        __LOOP_STATE__.LOOP.time
+    except RuntimeError:
+        return False
+    else:
+        return True
+
+
+V = TypeVar('V')
+
+
+class Environment:
+    """
```

Comment:
Needed to allow ``\ `` escape.

```suggestion
    r"""
```